src/core.jl

```diff
@@ -143,10 +143,10 @@ A[ClosedInterval(0.,.3), [:a, :c]]   # select an interval and two columns
 ```
 
 """
-struct AxisArray{T,N,D,Ax} <: AbstractArray{T,N}
+struct AxisArray{T,N,D<:AbstractArray{T,N},Ax<:NTuple{N,Axis}} <: AbstractArray{T,N}
```

Comment:
Long time ago there were some very subtle and surprising dispatch issues in doing this.  Looks like at least the minimal case I reported still exists:

```
julia> struct Foo{A<:Real}; end
       f(x::Foo{A}) where A = 1
       f(x::Foo, y...) = 2
f (generic function with 2 methods)

julia> f(Foo{Int}())
2
```

Not sure if any of the method sorting still depends on that behavior.

Reply:
Ah, the ambiguity no longer occurs but that still happens. 

Do you know if there is a Julia issue for this behaviour?

Reply:
https://github.com/JuliaLang/julia/issues/6383#issuecomment-75011574